### ISSUE TITLE
Once setting is configured with an env var, it only be changed with env var and configuring setting env var to be "" string changes value in rancher api but not in kubernetes

### ISSUE THREAD

**Issue by rmweir** [rancher/rancher#37998]
**Rancher Server Setup**
- Rancher version: v2.5.12

**Information about the Cluster**
- Kubernetes version: k3s 1.19 but probably doesn't matter
   
**Describe the bug**

~~Unsetting CATTLE_AGENT_IMAGE before an upgrade does not cause downstream cluster agents to use proper image. They will continue using the value that was previous set for CATTLE_AGENT_IMAGE. This is happening because setting the env var properly sets the setting in rancher. You can check this by running `kubectl get setting agent-image`.~~
The above works as intended but a related issue was found that will be tracked using this issue.

When a setting was configured using an env var of the format
CATTLE_SETTING_NAME, the source was set to "env". That source
would then persist even when the env var was unset in the future.
This prevented users from ever changing a setting value through the
rancher API or UI.

Also, when using env var to set the settings value to empty string, the rancher api will return empty string but the value is not updated in kubernetes. This will cause a change in behavior once the setting is interacted with through means other than norman. This could be moved to webhook but that won't be necessary once this is fixed. Once validation is moved to webhook, kubectl workaround won't be an option either.

There is a work around. You can use `kubectl edit setting <setting-name>`, then set the value and source fields to "" (empty string). Then redeploy rancher. You can also set the value to empty string with kubectl as well.

**To Reproduce**
1. Setup rancher installation.
2. Restart/upgrade rancher using env var such as CATTLE_EKS_REFRESH_CRON to set a setting that is present in the setting UI.
3. Try to change the value in the UI, should not be possible. The option will not be available in UI but even the API should not work for this.

for empty val scenario:
1. Set settings value to something that is a non empty string. You can use env var, rancher, or kubectl.
2. Now set it to empty string, "", using CATTLE_<setting-name> env var.
3. Get the setting from rancher will show the value as "" but if you get the value with kubectl it will be the last non empty string value you set it as.

**Result**
~~Downstream cluster agents are still using rancher/rancher-agent:v2.5.12.~~
The option will not be available in UI but even the API should not work for this.

**Expected Result**

The value of the setting should be the same but you should now be able to change it using the API and UI. Setting the env var value to empty string should now be reflected in kubernetes and in rancher api.

**Comment by cbron** [comment]
Moving this back to 2.7.1 now

**Comment by brudnak** [comment]
# 🚧 REOPENED

>TLDR: See the last test case in the table at the very bottom
> This was about the test case: Try to change the value in the UI, should not be possible. The option will not be available in UI but even the API should not work for this.

## Reproduction Environment


| Component                        | Version / Type                          |
| -------------------------------- | --------------------------------------- |
| Rancher version                  | v2.7.0                                  |
| Installation option              | Helm (high availability)                |
| RKE binary version used          | v1.4.0                                  |
| If Helm Chart k8s cluster        | v1.24.6                                 |
| Cert Details                     | Let's Encrypt / `nginx`                 |
| Docker version                   | 20.10.7, build f0df350                  |
| Helm version                     | v2.16.8-rancher2                        |
| Downstream cluster type          | not applicable                          |
| Downstream K8s version           | not applicable                          |
| Authentication providers enabled | local                                   |
| Logged in user role              | Administrator                           |
| Browser type                     | Google Chrome                           |
| Browser version                  | 108.0.5359.98 (Official Build) (x86_64) |


<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.7.1/cert-manager.crds.yaml
```

```shell
helm install cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --create-namespace \
  --version v1.7.1
```

```shell
helm install rancher rancher-latest/rancher \
  --namespace cattle-system \
  --set hostname=$URL_VAR \
  --set bootstrapPassword=$PASSWORD_VAR \
  --set ingress.tls.source=letsEncrypt \
  --set letsEncrypt.email=$EMAIL_VAR \
  --set letsEncrypt.ingress.class=nginx \
  --set rancherImageTag=v2.6.9 \
  --version 2.6.9 \
  --set extraEnv\[0\].name="CATTLE_PASSWORD_MIN_LENGTH" --set-string extraEnv\[0\].value="15"
```

</details>


## Reproduction steps

1. Install Rancher using helm on top of an RKE1 cluster
2. Provide `--set extraEnv\[0\].name="CATTLE_PASSWORD_MIN_LENGTH" --set-string extraEnv\[0\].value="15"` to the helm install command
3. Observe in the Rancher UI that the password minimum length is not changeable
   1. Starting from the **default** Rancher homepage `/dashboard/home`
   2. Click `hamburger menu` >>> `Global Settings` >>> scroll down and see the option for `password-min-length`
   3. Notice that it's not editable via the UI (there is no `three dots / ellipsis` to edit it with)
4. Trying un-setting the env var by restarting Rancher without the env var
5. Verify that the env var is not shown in the Rancher deployment
   1. Starting from the **default** Rancher homepage `/dashboard/home`
   2. Click `hamburger menu` >>> `local` >>> `Workload` >>> `Deployments` >>> filtered to `All Namespaces` >>> right click the `rancher` deployment under the `cattle-system` namespace >>> `Edit Config` >>> check under `Environment Variables` >>> and see that the password min length was removed
6. Now check the `Global Settings` again, and you are still not able to edit the password min length in the UI

## Additional Info

## RESULTS

#### ✅ Expected

To be able to change the password minimum length in the Global Settings UI after un-setting the `CATTLE` env var that set it

#### ❌ Actual

Was unable to change the minimum password length in the UI after un-setting that env var

<hr>

## Validation Environment


| Component                        | Version / Type                                                                     |
| -------------------------------- | ---------------------------------------------------------------------------------- |
| Rancher version                  | v2.7-830fa58ccecfb8915fc54862a9b19fdfeb2c0f8c-head                                 |
| Rancher commit link              | https://github.com/rancher/rancher/commit/830fa58ccecfb8915fc54862a9b19fdfeb2c0f8c |
| Installation option              | Helm (high availability)                                                           |
| RKE binary version used          | v1.4.0                                                                             |
| If Helm Chart k8s cluster        | v1.24.6                                                                            |
| Cert Details                     | Let's Encrypt / `nginx`                                                            |
| Docker version                   | 20.10.7, build f0df350                                                             |
| Helm version                     | v2.16.8-rancher2                                                                   |
| Downstream cluster type          | not applicable                                                                     |
| Downstream K8s version           | not applicable                                                                     |
| Authentication providers enabled | local                                                                              |
| Logged in user role              | Administrator                                                                      |
| Browser type                     | Google Chrome                                                                      |
| Browser version                  | 108.0.5359.98 (Official Build) (x86_64)                                            |


<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.7.1/cert-manager.crds.yaml
```

```shell
helm install cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --create-namespace \
  --version v1.7.1
```

```shell
helm install rancher rancher-latest/rancher \
  --namespace cattle-system \
  --set hostname=$URL_VAR \
  --set bootstrapPassword=$PASSWORD_VAR \
  --set ingress.tls.source=letsEncrypt \
  --set letsEncrypt.email=$EMAIL_VAR \
  --set letsEncrypt.ingress.class=nginx \
  --set rancherImageTag=v2.7.-head \
  --version 2.7.0 \
  --set extraEnv\[0\].name="CATTLE_PASSWORD_MIN_LENGTH" --set-string extraEnv\[0\].value="15"
```

</details>

## Validation steps

1. Install Rancher using helm on top of an RKE1 cluster
2. Provide `--set extraEnv\[0\].name="CATTLE_PASSWORD_MIN_LENGTH" --set-string extraEnv\[0\].value="15"` to the helm install command
3. Observe in the Rancher UI that the password minimum length is not changeable
   1. Starting from the **default** Rancher homepage `/dashboard/home`
   2. Click `hamburger menu` >>> `Global Settings` >>> scroll down and see the option for `password-min-length`
   3. Notice that it's not editable via the UI (there is no `three dots / ellipsis` to edit it with)
4. Trying un-setting the env var by restarting Rancher without the env var
5. After un-setting the env var you can now change the password min length via the UI in Global Settings 

## Additional Info

## RESULTS

#### ✅ Expected

To be able to edit the password min length after un-setting the env var

#### ✅ Actual

Was able to edit the password min length after un-setting the env var


## 📋 QA Test Plan

<table>
    <tbody>
        <tr>
            <th>Scenario</th>
            <th>Steps</th>
            <th>Expected</th>
            <th>Actual & Pass/Fail</th>
        </tr>
        <tr>
            <td>After un-setting the env var. Update the password min length and verify that the new value is being utilized</td>
            <td>
                <ol>
                    <li>Create a Rancher 2.7.0 with a password min length env var set to 15</li>
                    <li>Upgrade to Rancher 2.7-head</li>
                    <li>After upgrade to 2.7-head unset the env var</li>
                    <li>Now change the password min length from 15 to 16 using the Rancher UI in the Global Settings section</li>
                    <li>After changing the min password length to 16 attempt to create a new standard user with a 15 character password</li>
                </ol>
            </td>
            <td>Expected: For the changed Global Setting after un-setting an env var to be applied</td>
            <td>Actual: The change from 15 to 16 password min length was successfully applied after un-setting<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Start with Rancher 2.7.0 with an env var, then upgrade to 2.7-head and unset the env var</td>
            <td>
                <ol>
                    <li>Start Rancher 2.7.0 with CATTLE_PASSWORD_MIN_LENGTH env var set</li>
                    <li>Upgrade Rancher to 2.7-head</li>
                    <li>After upgrading unset the env var</li>
                    <li>Verify that the password min length is able to be changed via the UI</li>
                </ol>
            </td>
            <td>Expected: To be able to change the password min length after un-setting the env var</td>
            <td>Actual: Was able to change the min password length after un-setting the nev var<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Start with Rancher in 2.7.0 create an env var, attempt to unset it in 2.7.0, then upgrade to 2.7-head</td>
            <td>
                <ol>
                    <li>Start with a Rancher 2.7.0 with an env var set CATTLE_PASSWORD_MIN_LENGTH</li>
                    <li>Unset the env var while staying on 2.7.0</li>
                    <li>The min password is still not changeable in the UI</li>
                    <li>Then after un-setting, upgrade Rancher to 2.7-head</li>
                    <li>Check the Global Settings and attempt to change min password in the UI</li>
                </ol>
            </td>
            <td>Expected: After upgrading Rancher I expected to be able to change the min password length using the UI</td>
            <td>Actual: I was able to change the min password length after upgrade<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Unsure tha the UI option <code>Use the default value</code> still works after un-setting the env var</td>
            <td>
                <ol>
                    <li>Stat Rancher 2.7-head with the CATTLE_PASSWORD_MIN_LENGTH env var set</li>
                    <li>Unset the env var</li>
                    <li>Go to Global Settings</li>
                    <li>For the setting password-min-length attempt to edit and use the button <code>Use the default value</code> to make sure it still works</li>
                </ol>
            </td>
            <td>Expected: To be able to use the UI to revert to the default value of 12</td>
            <td>Actual: I was able to reset the default value to 12 using the UI<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Try to change the value in the UI, should not be possible. The option will not be available in UI but even the API should not work for this.</td>
            <td>
                <ol>
                    <li>Pulled this test case from: https://github.com/rancher/rancher/pull/39453#issue-1427929273</li>
                    <li>Create a Rancher 2.7-head with CATTLE_PASSWORD_MIN_LENGTH set to 13</li>
                    <li>Attempt to change the password min length with the API using Postman</li>
                    <li>Use a <code>GET</code> request to the following</li>
                    <li><code>/v1/management.cattle.io.settings/password-min-length</code></li>
                    <li>Look for the password-min-length setting and grab the <code>resourceVersion</code> to use for example: 57929</li>
                    <li>Also grab the <code>uid</code></li>
                    <li>Now Create a <code>PUT</code> request to the following </li>
                    <li><code>v1/management.cattle.io.settings/password-min-length</code></li>
                    <li>in the PUT request change the password min length from 13 to 20</li>
                    <li>The PUT request is successful <code>200</code></li>
                    <li>The Rancher Global Settings UI is updated and shows the password-min-length as 20</li>
                    <li>It is now also editable in the Rancher UI</li>
                    <li>Try creating a new user with a password that is 12 characters</li>
                    <li>Rancher isn't detecting what is shown in the Global Settings UI as 20 characters for the min password length</li>
                    <li>Even though the API allowed to change and gave status 200 the min password length is still 13</li>
                    <li>See the attached screenshot below this table</li>
                </ol>
            </td>
            <td>Expected: To not be able to change the min password length when it's blocked from the UI and set via an env var</td>
            <td>Actual: Was able to "change" the password min length via the API when it was being blocked in the UI. This broke a couple things. It showed the three dots/ellipsis in the Global Settings to change the min password length, when it was previously not an option. The updated value from 13 to 20 was also not being enforced by Rancher and I was able to create users with min passwords less than 20 characters<br><br>Pass/Fail: ❌<code>FAIL</code></td>
        </tr>
    </tbody>
</table>


![2022-12-15_18-08-41](https://user-images.githubusercontent.com/40877149/208001623-31fc3cdd-9790-4b74-9f0f-948a1587eab2.png)

**Comment by brudnak** [comment]
The developer team with QA has determined that this issue will only be validated through the user interface. A separate issue will be created to handle API validation. The test case, `Try to change the value in the UI, should not be possible. The option will not be available in UI but even the API should not work for this,` should be checked against the new issue. The user interface validations have passed.
